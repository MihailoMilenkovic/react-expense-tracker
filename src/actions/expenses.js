import uuid from "uuid";
import database from "../firebase/firebase";
//without a database:
//component calls action generator
//action generator returns object
//component dispatches object
//redux store changes

//with a database:
//component calls action generator
//action generator returns function
//component dispatches function
//function runs (can dispatch other actions and do other things)

export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    return database
      .ref("expenses")
      .push(expense)
      .then((ref) => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense,
          })
        );
      });
  };
};

//remove expense
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

//edit expense
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});
