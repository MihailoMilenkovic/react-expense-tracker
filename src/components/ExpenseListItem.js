import React from "react";
import {Link} from "react-router-dom";
//desc, amount. createdAt
const ExpenseListItem = ({id,description,amount,createdAt}) => (
    <div>
        <h3>{description}</h3>
        <p>{amount} - {createdAt}</p>
        <Link to={`/edit/${id}`}>edit expense</Link>
    </div>
);

export default ExpenseListItem;