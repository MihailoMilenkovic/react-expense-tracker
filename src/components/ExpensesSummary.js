import React from "react";
import numeral from "numeral";
import ExpensesTotal from "../selectors/expenses-total";
import { connect } from "react-redux";
import selectExpenses from "../selectors/expenses";

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = expenseCount === 1 ? "expense" : "expenses";
  const formatedExpensesTotal = numeral(expensesTotal / 100).format("$0,0.00");
  return (
    <div>
      <h1>
        Viewing {expenseCount} {expenseWord} totaling {formatedExpensesTotal}
      </h1>
    </div>
  );
};

const mapStateToProps = (state) => {
  const expenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: expenses.length,
    expensesTotal: ExpensesTotal(expenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);