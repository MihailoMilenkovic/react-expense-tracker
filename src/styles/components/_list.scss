.list-header {
  background: $light-grey;
  border: 1px solid darken($color: $light-grey, $amount: 7%);
  color: grey;
  display: flex;
  justify-content: space-between;
  padding: $s-size $m-size;
}

.list-item {
  border: 1px solid darken($color: $light-grey, $amount: 7%);
  border-top: none;
  display: flex;
  flex-direction: column;
  color: $dark-grey;
  padding: $s-size;
  text-decoration: none;
  transition: background 0.3s ease;
  &:hover {
    background: $light-grey;
  }

  @media (min-width: $desktop-breakpoint) {
    align-items: center;
    flex-direction: row;
    justify-content: space-between;
    padding: $m-size;
  }
}
.list-body {
  margin-bottom: $m-size;
  @media (min-width: $desktop-breakpoint) {
    margin-bottom: $l-size;
  }
}

.list-item--message {
  align-items: center;
  justify-content: center;
  color: $grey;
  padding: $m-size;
  &:hover {
    background: none;
  }
}

.list-item__title {
  margin: 0;
  word-break: break-all;
}

.list-item__subtitle {
  color: $grey;
  font-size: 1.4rem;
}

.list-item__data {
  margin: $s-size 0 0 0;
  @media (min-width: $desktop-breakpoint) {
    margin: 0;
    padding-left: $s-size;
  }
}
