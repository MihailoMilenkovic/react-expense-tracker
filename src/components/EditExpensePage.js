import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(expense.id, expense);
    this.props.history.push("/");
  };
  onRemove = () => {
    this.props.removeExpense({ d: this.props.expense.id });
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onRemove}>remove</button>
      </div>
    );
  }
}
const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      (expense) => expense.id === props.match.params.id
    ),
  };
};
const mapDispatchToProps = (dispatch, props) => {
  return {
    editExpense: (expId, expense) => dispatch(editExpense(expId, expense)),
    removeExpense: (data) => dispatch(removeExpense(data)),
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
