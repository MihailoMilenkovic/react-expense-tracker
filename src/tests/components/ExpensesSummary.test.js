import React from "react";
import { ExpensesSummary } from "../../components/ExpensesSummary";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses";

test("expense summary should be displayed correctly for list with one expense", () => {
  const wrapper = shallow(<ExpensesSummary
    expenseCount={1}
    expensesTotal={430}
  />);
  expect(wrapper).toMatchSnapshot();
});
test("expense summary should be displayed correctly for list with multiple expenses", () => {
  const wrapper = shallow(<ExpensesSummary
    expenseCount={2}
    expensesTotal={430}
  />);
  expect(wrapper).toMatchSnapshot();
});