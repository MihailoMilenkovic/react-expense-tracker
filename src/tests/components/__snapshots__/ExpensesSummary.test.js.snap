// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expense summary should be displayed correctly for list with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    2
     
    expenses
     totaling 
    $4.30
  </h1>
</div>
`;

exports[`expense summary should be displayed correctly for list with one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totaling 
    $4.30
  </h1>
</div>
`;
